source/dlangbot/database.d

```diff
@@ -0,0 +1,109 @@
+/**
+Implements a database to keep track of the Points scored
+by Authors that fix bugs. The table has the following
+columns:
+
+Github account (Primary key)
+Total Points
+*/
+module dlangbot.database;
+
+import ae.sys.database : Database;
+
+private Database db;
+private string[] severities = ["enhancement", "trivial", "minor",
+            "normal", "major", "critical", "blocker", "regression"];
+enum table_schema = q"SQL
+CREATE TABLE [LeaderBoard] (
+[Id]  INTEGER PRIMARY KEY NOT NULL,
+[Name] TEXT NOT NULL,
+[Enhancement] INTEGER,
+[Trivial] INTEGER,
+[Minor] INTEGER,
+[Normal] INTEGER,
+[Major] INTEGER,
+[Critical] INTEGER,
+[Blocker] INTEGER,
+[Regression] INTEGER,
+[Points] INTEGER
+);
+SQL";
+
+/**
+Checks whether a database file already exists.
+If that is the case, it returns the database,
+otherwise it creates it.
+*/
+Database initDatabase(string databasePath)
+{
+    import std.file : exists;
+    import ae.sys.file : ensurePathExists;
+
+    version(unittest)
+    {
+        if (databasePath == ":memory:")
+            return Database(databasePath, [table_schema,]);
+    }
+
+    if (databasePath.exists)
+        return Database(databasePath);
+
+    ensurePathExists(databasePath);
+    return Database(databasePath, [table_schema,]);
+}
+
+Database getDatabase()
+{
+    return db;
+}
+
+void updateScoreBoard(string name, ulong id, uint[string] groupedSevs)
+{
+    // association between points received and severity
+    uint[string] pointsPerSev = ["enhancement" : 10, "trivial" : 10,
+            "minor" : 15, "normal" : 20, "major" : 50, "critical" : 75,
+            "blocker" : 75, "regression" : 100];
+
+    // compute total numner of points received
+    uint points;
+    foreach(sev; severities)
+    {
+        if (sev !in groupedSevs)
+        {
+            groupedSevs[sev] = 0;
+        }
+        else
+        {
+            points += groupedSevs[sev] * pointsPerSev[sev];
+        }
+    }
+    db.stmt!"
+    INSERT OR REPLACE INTO [LeaderBoard] ([Id], [Name], [Enhancement], [Trivial],
+                [Minor], [Normal], [Major], [Critical], [Blocker], [Regression], [Points])
+    VALUES (?1, ?2, COALESCE((SELECT [Enhancement] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?3,
+                COALESCE((SELECT [Trivial] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?4,
+                COALESCE((SELECT [Minor] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?5,
+                COALESCE((SELECT [Normal] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?6,
+                COALESCE((SELECT [Major] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?7,
+                COALESCE((SELECT [Critical] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?8,
+                COALESCE((SELECT [Blocker] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?9,
+                COALESCE((SELECT [Regression] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?10,
+                COALESCE((SELECT [Points] FROM [LeaderBoard] WHERE [Id] = ?1), 0) + ?11
```

Comment:
This records both the issue count and the points? Though the latter could be calculated from the former? If this is by design then I think a comment would be nice to explain why it was done this way.

Reply:
It's easier (from an implementation perspective) to keep the total number of points and incrementally add to it than to compute it every time. You can see it as an optimization at the expense of space.

Reply:
Also, at the beginning I wasn't entirely sure how the web page was going to obtain the points information. I initially envisioned it as the web page issuing a request to obtain the points (from this perspective it makes sense to simply store the points - it eases the implementation on the web page side). However, looking at dlang-bot it seems more natural for the bot to simply issue an update request to the web page every time a bug is fixed. This could work without storing the total number of points, but would require an extra query to compute the points on the bot side.